feat: regex tester

shashankshet opened the pull request:
Hi Team,

I've added a new regex testing feature that lets users input a regex pattern and a test string to see if they match. It includes error handling for invalid patterns.

If you find this feature useful, please consider merging it into the main branch.

Thanks!

danigrant commented:
That's such an awesome tool to add!! Thank you so much @shashankshet! 

Would love to send you a swag thanks :) if you send me an email (dani@jam.dev) I'll get your mailing details and send something over! 😃

shashankshet commented:
Thanks! @danigrant I’m glad you liked the tool. I appreciate the offer and will send you an email with my mailing details right away.

peckz commented on pages/utilities/regex-tester.tsx:
```diff
@@ -0,0 +1,96 @@
+    []
+  );
+
+  const handleTest = useCallback(() => {
+    try {
+      const regex = new RegExp(pattern);
+      const match = regex.test(testString);
+      setResult(match ? "Match found" : "No match found");
+    } catch (error) {
+      setResult("Invalid regex pattern");
+    }
+  }, [pattern, testString]);
```
Since this function covers very basic use case, my suggestion is to extend this function and allow users to pass full regex patterns with flags. With this approach we will have better error handling / error messaging.

First, let's add function outside of the react component, to the bottom of the file and name it `createRegex()`

```
const createRegex = (pattern: string): RegExp => {
  if (!pattern) {
    throw new Error("Pattern cannot be empty");
  }

  if (pattern.startsWith("/") && pattern.lastIndexOf("/") !== 0) {
    const lastSlashIndex = pattern.lastIndexOf("/");
    const patternBody = pattern.slice(1, lastSlashIndex);
    const flags = pattern.slice(lastSlashIndex + 1);

    if (!/^[gimsuy]*$/.test(flags)) {
      throw new Error("Invalid regex flags");
    }

    return new RegExp(patternBody, flags);
  } else {
    return new RegExp(pattern);
  }
};
```

Then utilize function in `handleTest()`, I think it makes sense to add matches as part of `result`
For better error handling, let's use `error.message` as output instead of generic `Invalid regex pattern`

```
const handleTest = useCallback(() => {
  try {
    const regex = createRegex(pattern);
    const matches = testString.match(regex);

    if (matches) {
      setResult(`Match found:\n${matches.join("\n")}`);
    } else {
      setResult("No match found");
    }
  } catch (error) {
    if (error instanceof Error) {
      setResult(error.message);
    }
  }
}, [pattern, testString]);
```

peckz commented on pages/utilities/regex-tester.tsx:
```diff
@@ -0,0 +1,96 @@
+            <Label>Regex Pattern</Label>
+            <Textarea
+              rows={4}
+              placeholder="Enter regex pattern here"
```
Let's be more specific with example, like `Enter regex pattern here (e.g., /pattern/g)`

peckz commented:
@shashankshet Thank you for the initiative, it's an amazing idea! Once you implement suggested changes, I'm going to design component that will highlight matched characters, similar to https://regexr.com/. Once that's done, we will be ready to ship this tool!

shashankshet commented:
Hi @peckz,

I've addressed the review comments:

1. Refactored the handleTest function to use the new createRegex() function, allowing full regex patterns with flags. This also improves error handling and messaging.
2. Updated the placeholder text to be more specific, showing an example regex pattern.
3. Additionally, I enhanced the UI by displaying the match status ("Match found" or "No match found") next to the "Result" label. The status is now color-coded, with "Match found" appearing in green and "No match found" in red for better visual feedback.

Let me know if there's anything else!

peckz commented:
@shashankshet Thank you! It's my turn, stay tuned!

peckz commented:
@shashankshet Thanks for the great idea and initial setup! 
I added text highlighter and set of tests to cover `createRegex()`.
Welcome to the contributors list!
